test/test_python_reproducibility.py

```diff
@@ -1,265 +1,99 @@
+        ('INa_block', {'scaled_APD90': 1, 'detailed_voltage': 2}),
+        ('INa_IV_curves', {'normalised_peak_currents': 1, 'current_activation': 2}),
+        ('NCX_block', {'scaled_resting_potential': 1, 'scaled_APD90': 1, 'detailed_voltage': 2}),
+        ('RyR_block', {'scaled_APD90': 1, 'detailed_voltage': 2}),
+        ('S1S2', {'S1S2_slope': 1}),
+        ('SteadyStateRestitution', {'APD': 2, 'restitution_slope': 1}),
+        ('SteadyStateRunner', {'num_paces': 0, 'detailed_voltage': 1}),
+        ('SteadyStateRunner4Hz', {'num_paces': 0, 'detailed_voltage': 1}),
+    ],
+    ids=proto_id_fn
+)
+def test_fc_experiment(model_name, protocol_settings, tmpdir, request):
+    """Test reproducibility of an experiment, i.e. application of a protocol to a model.
+
+    :param model_name: name of model to run, i.e. no path or extension
+    :param protocol_settings: pair with details of a protocol to run and expected results:
+        first element gives the name of protocol to run, i.e. no path or extension;
+        second is a dictionary of outputs to check against reference data,
         mapping output name to number of dimensions
-    :returns: a tuple (modelName, protoName, boolean result, output text, error/warning messages)
     """
-    messages = []
-    result = True
-    output = StringIO()
-    proto = None
-    with RedirectStdStreams(output, output):
-        try:
-            print("Applying", protoName, "to", modelName,
-                  "on process", test_support.get_process_number(), "of", CHASTE_NUM_PROCS)
-            setproctitle(
-                'python worker %d running %s on %s' % (
-                    test_support.get_process_number(), protoName, modelName))
-            proto = fc.Protocol('protocols/%s.txt' % protoName)
-            proto.set_output_folder(os.path.join(CHASTE_TEST_OUTPUT, 'Py_FunctionalCuration', modelName, protoName))
-            proto.set_model('cellml/%s.cellml' % modelName)
-            for input in ['max_paces', 'max_steady_state_beats']:
-                try:
-                    proto.set_input(input, fc.language.values.Simple(1000))
-                except Exception:
-                    pass  # Input doesn't exist
-            proto.run()
-        except Exception:
-            result = False
-            messages.append(traceback.format_exc())
+    proto_name, expected_outputs = protocol_settings
+    print(f'Applying {proto_name} to {model_name}')
+
+    # If there are missing reference results then this is expected to fail
+    data_folder = f'test/output/real/{model_name}/{proto_name}'
+    for name in expected_outputs:
+        if not os.path.exists(os.path.join(data_folder, 'outputs_' + name + '.csv')):
+            request.applymarker(pytest.mark.xfail(reason=f'Missing reference result {name}', strict=True))
+            break
+
+    # Try to run the protocol
+    proto = fc.Protocol('test/protocols/real/%s.txt' % proto_name)
+    os.environ['CHASTE_TEST_OUTPUT'] = str(tmpdir)
```

Comment:
Are we using `CHASTE_TEST_OUTPUT` anywhere outside of `fc` now? Because if so we could probably communicate it in some better way, e.g. a global variable `fc.TEST_OUTPUT`?

Reply:
That's a good idea. I think we already have an issue to get rid of the env var?

Reply:
Yeah in https://github.com/ModellingWebLab/weblab-fc/issues/59. Linking it here for future ref!